File: app/src/test/java/com/weiqi/sgf/SgfTest.kt
package com.weiqi.sgf

import com.weiqi.engine.GameConfig
import com.weiqi.engine.GameState
import com.weiqi.engine.MoveIntent
import com.weiqi.engine.MoveResult
import com.weiqi.engine.MoveType
import com.weiqi.engine.Point
import com.weiqi.engine.Rules
import java.time.LocalDate
import org.junit.Assert.assertTrue
import org.junit.Test

class SgfTest {

    @Test fun exportsBasicGame() {
        var s = GameState.newGame(GameConfig(boardSize = 9, komi = 7.5))
        s = (Rules.apply(s, MoveIntent(MoveType.PLACE_STONE, Point(2, 3))) as MoveResult.Accepted).newState
        s = (Rules.apply(s, MoveIntent(MoveType.PASS)) as MoveResult.Accepted).newState
        val sgf = Sgf.export(s, score = null, blackName = "B", whiteName = "W", date = LocalDate.of(2026, 5, 3))
        assertTrue(sgf.startsWith("(;GM[1]FF[4]"))
        assertTrue(sgf.contains("SZ[9]"))
        assertTrue(sgf.contains("KM[7.5]"))
        assertTrue(sgf.contains("RU[Chinese]"))
        assertTrue(sgf.contains(";B[dc]"))   // (row=2,col=3) -> col 'd', row 'c'
        assertTrue(sgf.contains(";W[]"))      // pass
        assertTrue(sgf.endsWith(")"))
    }
}

File: test/sgf/sgf_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:go_game/src/domain/game_state.dart';
import 'package:go_game/src/domain/models.dart';
import 'package:go_game/src/domain/rules.dart';
import 'package:go_game/src/sgf/sgf.dart';

void main() {
  test('exports basic game', () {
    var s = GameState.newGame(const GameConfig(boardSize: 9, komi: 7.5));
    s = Rules.apply(s, const MoveIntent.place(Point(2, 3)))
        .newStateAs<GameState>();
    s = Rules.apply(s, const MoveIntent.pass()).newStateAs<GameState>();
    final sgf = Sgf.export(
      s,
      blackName: 'B',
      whiteName: 'W',
      date: DateTime(2026, 5, 3),
    );
    expect(sgf.startsWith('(;GM[1]FF[4]'), isTrue);
    expect(sgf.contains('SZ[9]'), isTrue);
    expect(sgf.contains('KM[7.5]'), isTrue);
    expect(sgf.contains('RU[Chinese]'), isTrue);
    expect(sgf.contains('GC[Chinese rules'), isTrue);
    expect(sgf.contains(';B[dc]'), isTrue); // (row=2,col=3) -> col 'd', row 'c'
    expect(sgf.contains(';W[]'), isTrue);
    expect(sgf.endsWith(')'), isTrue);
  });

  test('emits matching RU and GC for Japanese rules', () {
    final s = GameState.newGame(const GameConfig(
      boardSize: 9,
      ruleset: Ruleset.japanese,
      komi: 6.5,
      superkoMode: SuperkoMode.none,
    ));
    final sgf = Sgf.export(s, date: DateTime(2026, 5, 11));
    expect(sgf.contains('RU[Japanese]'), isTrue);
    expect(sgf.contains('GC[Japanese rules'), isTrue);
    expect(sgf.contains('territory + prisoners'), isTrue);
    expect(sgf.contains('basic ko'), isTrue);
  });
}
